plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/changedList/ChangedListViewImpl.java

```diff
@@ -172,9 +203,105 @@ public void onClick(ClickEvent event) {
+        List<String> allPaths = new ArrayList<>();
+        for (String file : allFiles) {
+            String path = file.substring(0, file.lastIndexOf("/"));
+            if (!allPaths.contains(path)) {
+                allPaths.add(path);
+            }
+        }
+        String commonPath = getCommonPath(allPaths);
+        Map<String, Node> preparedNodes = new HashMap<>();
+        for (int i = getMaxNestedLevel(allFiles); i > 0; i--) {
+            Map<String, List<Node>> currentChildNodes = new HashMap<>();
+            for (String file : allFiles) {
+                if (file.split("/").length != i) {
+                    continue;
+                }
+                String path = file.substring(0, file.lastIndexOf("/"));
+                String name = file.substring(file.lastIndexOf("/") + 1);
+                String pathName = path.isEmpty() ? name : path + "/" + name;
+                Node fileNode = new ChangedFileNode(pathName, items.get(pathName), nodesResources, true) {
+                    @Override
+                    public void actionPerformed() {
+                        delegate.onCompareProvided();
+                    }
+                };
+                if (currentChildNodes.keySet().contains(path)) {
+                    currentChildNodes.get(path).add(fileNode);
+                } else {
+                    List<Node> listFiles = new ArrayList<>();
+                    listFiles.add(fileNode);
+                    currentChildNodes.put(path, listFiles);
+                }
+            }
+            for (String path : currentChildNodes.keySet()) {
+                Node folder = new ChangedFolderNode(getFolders(allPaths, path), nodesResources);
+                folder.setChildren(currentChildNodes.get(path));
+                preparedNodes.put(path, folder);
+            }
+            List<String> currentPaths = new ArrayList<>(preparedNodes.keySet());
+            for (String currentPath : currentPaths) {
+                List<Node> nodesToNest = new ArrayList<>();
+                for (String nestedItem : currentPaths) {
+                    if (!currentPath.equals(nestedItem) && nestedItem.startsWith(currentPath)) {
+                        nodesToNest.add(preparedNodes.remove(nestedItem));
+                    }
+                }
+                if (nodesToNest.isEmpty()) {
+                    continue;
+                }
+                Collections.sort(nodesToNest, new NameComparator());
+                nodesToNest.addAll(currentChildNodes.get(currentPath));
+                if (currentPath.equals(commonPath)) {
+                    return nodesToNest;
+                } else {
+                    preparedNodes.get(currentPath).setChildren(nodesToNest);
+                }
+            }
+        }
+        return new ArrayList<>(preparedNodes.values());
+    }
+
+    private String getFolders(List<String> allPaths, String comparedPath) {
+        String[] segments = comparedPath.split("/");
```

Comment:
Use `org.eclipse.che.ide.resource.Path` it will help you to reduce the code size to work with path segments.